sdk/core/Azure.Core/src/Pipeline/UserAgentValue.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace Azure.Core.Pipeline
+{
+    /// <summary>
+    /// Information about the package to be included in UserAgent telemetry
+    /// </summary>
+    public class UserAgentValue
+    {
+        private string _userAgent;
+
+        /// <summary>
+        /// Initialize an instance of <see cref="UserAgentValue"/> by extracting the name and version information from the <see cref="Assembly"/> associated with the <paramref name="type"/>.
+        /// </summary>
+        /// <param name="type">The <see cref="Type"/> used to generate the package name and version information for the <see cref="UserAgentValue"/> value.</param>
+        /// <param name="applicationId">An optional value to be prepended to the <see cref="UserAgentValue"/>.
+        /// This value overrides the behavior of the <see cref="DiagnosticsOptions.ApplicationId"/> property for the <see cref="HttpMessage"/> it is applied to.</param>
+        public UserAgentValue(Type type, string? applicationId = null)
+        {
+            var assembly = Assembly.GetAssembly(type);
+            if (assembly == null) throw new ArgumentException($"The type parameter {type.FullName} does not have a valid Assembly");
+            _userAgent = GenerateUserAgentString(assembly, applicationId);
```

Comment:
Just realized this will always generate this string, and if we are setting this per message that would be a large perf hit.  Can we come up with a way to either cache this or allow us to pass in the full value?

Reply:
The idea was that you create one (RP-shared) instance of UserAgentValue and then use it in every call. Is that not possible?

Reply:
What is happening here is that the applicationId is a mechanism that we expose to 3rd party libraries so that they can prepend their own user agents on top of the sdk library user agent.  For instance powershell will add things in there so that they get their own telemetry on top of ours.  The SDK user agent is constructed each time on line 27 based on the type that is used which will happen for 100% of the calls.  We either need another signature to accept the full string to add so that we can construct this once in our RP and simply pass it in, OR if we want to keep all of the logic on how to create this string inside Azure.Core we should have a way to cache this value by Type / applicationId tuple so if I can with the same value twice its simply reused and not reconstructed each time.

Reply:
Could you clarify why is the type being created on every call? Wouldn't this be created during the initialization of the pipeline and not change after that?

Reply:
I'm not sure I follow why this is a problem either. As @KrzysztofCwalina points out, the usage scenario that ResourceManager uses (with this new implementation swapped in) is illustrated in the PR summary. It shows the sub-client initialization, which creates the instance of UserAgentValue for that RP including the `ApplicationId` that the user passed to their client options. This value is cached for all calls made by the RP sub-client.

Reply:
This property is set on each message, each message is a new object instance, so each time we call this it will construct the string.